src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1027,16 +1028,13 @@ DatabaseShardImp::asyncFetch(
 
         auto const it{shards_.find(acquireIndex_)};
         if (it == shards_.end())
-            return false;
+            return true;
         shard = it->second;
     }
```

Comment:
This function doesn't look correct. We check if the `acquireIndex_` is in shards and we don't fetch if it's not? That doesn't make sense to me. I spoke with @miguelportilla and he agrees that this looks buggy. He suggested the following fix:
```
bool
DatabaseShardImp::asyncFetch(
    uint256 const& hash,
    std::uint32_t ledgerSeq,
    std::function<void(std::shared_ptr<NodeObject>&)>&& callback)
{
    auto const shardIndex{seqToShardIndex(ledgerSeq)};
    {
        std::lock_guard lock(mutex_);
        assert(init_);
        if (shards_.find(shardIndex) == shards_.end())
            return false;
    }

    // Otherwise post a read
    return Database::asyncFetch(hash, ledgerSeq, std::move(callback));
}
```
Notes:
* this is part of the patch here: https://github.com/seelabs/rippled/commit/f272b74741a8bc9d0e51f279fa978f9ed4a8ecf9
* this uses the new return value semantics (true means pending, false mean it's never coming).

Reply:
Actually, looking closer, `DatabaseShardImp::fetchNodeObject(` already does this check. This means we can get rid of all the overides. I'll make a patch to do so.

Reply:
Here's a second patch that builds on the first: https://github.com/ripple/rippled/commit/6147739d1a71b522597b632988fb055170c707dc @miguelportilla Please dummy check that it is OK to remove `DatabaseShardImp::asyncFetch` and use `fetchNodeObject` to do this check for us.

Reply:
The patches look good to me. The drawback is that we can now post a read knowing that the ledger sequence does not correspond to any shard. But it won't make it to IO as `DatabaseShardImp::fetchNodeObject` will return early (as you mentioned). This cleanup's benefit far outweighs that.